### ISSUE TITLE
 lambda layer using tf.image

### ISSUE THREAD

**Issue by katmatus** [Dobiasd/frugally-deep#144]
I am using Lambda layer which includes this function to extract patches in image

patch_one = tf.image.extract_glimpse(inputs[0], [26, 26], inputs[1][:, j, :], centered=False, normalized=False, noise='zero')

Is it possible to implement this custom layer in your library and load model?

**Comment by Dobiasd** [comment]
> Is it possible to implement this custom layer in your library and load model?

Probably, yes. :)
Please have a look at ["How to use lambda/custom layers?" in the FAQ](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-use-lambdacustom-layers).

**Comment by katmatus** [comment]
Hi, thank you for the answer. I followed you link, but I have some problems during compilation.
I created my class and function:
```
namespace fdeep 
{
	namespace internal
	{
		class LambdaPatches : public layer
		{
		public:
			explicit LambdaPatches(const std::string& name)
				: layer(name)
			{
			}

                        inline layer_ptr create_patches_layer(
				const get_param_f &, 
				const get_global_param_f &,
				const nlohmann::json &,
				const std::string &name)
                       {
                            return std::make_shared<LambdaPatches>(name);
                       }

                protected:
                       tensor5s apply_impl(const tensor5s &input) const override { return {sum_tensor5s(input)}; }
		};
	}
}
```

I try to load my model like this:

```
auto myFirstLayer = std::make_shared<fdeep::internal::LambdaPatches>("lambda_1");

auto model = fdeep::load_model("fdeep_model.json", true, fdeep::cout_logger,
                                   static_cast<fdeep::float_type>(0.00001),
                                   {
                                       {"lambda_1", myFirstLayer->create_patches_layer}
                                   });
```

And I have this errors:

> 'fdeep::internal::LambdaPatches::create_patches_layer': non-standard syntax; use '&' to create a pointer to member
> 
> 'fdeep::model fdeep::load_model(const std::string &,bool,const std::function<void (std::string)> &,fdeep::float_type,const fdeep::internal::layer_creators &)': cannot convert argument 5 from 'initializer list' to 'const fdeep::internal::layer_creators &'

What am I doing wrong?

**Comment by Dobiasd** [comment]
Your current version of `create_patches_layer` is a non-static member function of
class `LambdaPatches`, i.e., it requires an existing object of type `LambdaPatches` to be called (which would then be `this` in the function's body).

Syntactically, your current issue is creating a function reference to this, but even if you overcome this (with `std::bind(&LambdaPatches::create_patches_layer, ...`) the underlying issue would still exist.

A factory function is meant to be run without the object to be created already existing. :)

The FAQ links at this point to `create_add_layer`(https://github.com/Dobiasd/frugally-deep/blob/e3e1a6a2e011ef6255d6589a5ec0981c9d0ef1f9/include/fdeep/import_model.hpp#L594), which is a good example of a free function provided as a factory.

Your code distilled to a fixed minimal example, looks something like this:

```cpp
#include <functional>
#include <map>
#include <memory>
#include <string>

struct layer {
};

struct my_layer: layer {
};

layer create_patches_layer(int) {
    return my_layer();
}

using factory_function = std::function<layer(int)>;

using layer_creators = std::map<std::string, factory_function>;

void load_model(const layer_creators& custom_creators) {
}

int main() {
    const layer_creators creators = {{"my_layer", create_patches_layer}};
    load_model(creators);
}
```

So, basically, make `create_patches_layer` a free function, or a static member function.

In case you decide for the static member function, you can reference it (for conversion to an `std::function`) with `my_layer::create_patches_layer` instead of just `create_patches_layer` as in the free-function version.

**Comment by katmatus** [comment]
Thank you for quick answer, it seems to be loaded :). And my actual action for this layer is, that I have to ovveride `tensor5s apply_impl(const tensor5s &input) `and call my  C++ code which extract patches, is it right?

**Comment by Dobiasd** [comment]
Exactly. :+1:

**Comment by katmatus** [comment]
Hello Dobiasd, my lambda layer has output shape (1,52,26,26,1) , because of 52 patches. Then I apply conv2d and output is (1,1,26,26,32), but right output has to be (1,52,26,26,32). How to make it work properly?

**Comment by Dobiasd** [comment]
Can you provide a minimal example (code) that I can use to reproduce the problem?

**Comment by katmatus** [comment]
my lambda call, input shapes: (1,1,255,235,1), (1,1,1,52,2)
output shape: (1,52,26,26,1)
```
            tensor5s apply_impl(const tensor5s &input) const override
            {
                // convert tensor5 image to Eigen::Tensor and swap width and height
                std::vector<float> x_data = *input[0].as_vector();
                auto mapped_t = Eigen::TensorMap<Eigen::Tensor<float, 4, Eigen::RowMajor>>(
                    &x_data[0], 1, (int)input[0].height(), (int)input[0].width(), 1);

                Eigen::Tensor<float, 4> x_input = mapped_t.swap_layout();
             
                int num_points = input[1].width();
                tensor5s vec_patches;

                for (int i = 0; i < num_points; ++i)
                {
                    float offset_y = input[1].get(0, 0, 0, i, 0);
                    float offset_x = input[1].get(0, 0, 0, i, 1);

                    // convert tensor5 to std::vector<Eigen::IndexPair> offsets
                    std::vector<Eigen::IndexPair<float>> offset_vec = {{offset_x, offset_y}};

		    // extract patches around points
                    Eigen::GlimpseExtractionOp<int> patchesObj(settings::widthPatch, settings::heightPatch, offset_vec,
                                                           false, false, Eigen::ExtractGlimpsesNoiseMode::ZERO);

                    Eigen::DefaultDevice cpuDevice;
                    Eigen::Tensor<float, 4> x_output(1, settings::widthPatch, settings::heightPatch, 1);

                    patchesObj.eval<Eigen::Tensor<float, 4>, Eigen::Tensor<float, 4>, Eigen::DefaultDevice>(
                        x_input, x_output, cpuDevice);

		    // swap width and height
                    auto mapped_p = Eigen::TensorMap<Eigen::Tensor<float, 4, Eigen::RowMajor>>(
                        &x_output.data()[0], 1, settings::widthPatch, settings::heightPatch, 1);

					Eigen::Tensor<float, 4> x_swap = mapped_p.swap_layout();
                       
		    // convert Eigen::Tensor to tensor5
                    std::vector<float> vec_swap(x_swap.data(),x_swap.data() +
                                                                          settings::widthPatch * settings::heightPatch);

                    const fdeep::shared_float_vec sv(fplus::make_shared_ref<fdeep::float_vec>(vec_swap));
                    fdeep::tensor5 tensor_output(fdeep::shape5(1, 1, settings::heightPatch, settings::widthPatch, 1),
                                                 sv);

                    vec_patches.push_back(tensor_output);

                }

	        //tensor5s to tensor5
                tensor5 output_tensor = tensor5_from_depth_slices(vec_patches);

		// reshape tensor5 to (1,num_points,height,width,1)
		const fdeep::shared_float_vec reshape_pt( fplus::make_shared_ref<fdeep::float_vec> 
                (*output_tensor.as_vector()));

		fdeep::tensor5 tensor_output( fdeep::shape5(1, num_points, settings::heightPatch, 
                settings::widthPatch, 1), reshape_pt);

                return {tensor_output};
            }
```
after lambda layer goes in your conv_2d_layer with n_convolutions= 32, size = 3, stride =1
input shape is (1,52,26,26,1) and output shape is already  (1,1,26,26,32) ,I think it is because of convolution.hpp line 129, you create tensor with shape (1, 1, out_height, out_width, out_depth)

**Comment by Dobiasd** [comment]
Thanks. So to make sure I understand correctly: The problem you are describing is not related to the lambda layer, but to the implementation of the normal `Conv2D` layer in frugally deep, right?
Basically, you observe the following:

You have one `Conv2D(32)` layer taking input tensors of shape `(1,52,26,26,1)`.
In frugally-deep the output shape is `(1,1,26,26,32)`, however this does not match what Keras does.

Is the above correct?

**Comment by Dobiasd** [comment]
Assuming the above is right, I just tried to create such a minimal-example model in Keras:

```python3
# create_model.py
import numpy as np
from keras.layers import Input, Conv2D
from keras.models import Model

inputs = Input(shape=(52, 26, 26, 1))
x = Conv2D(32, (3, 3))(inputs)
model = Model(inputs=inputs, outputs=x)
model.compile(loss='mean_squared_error', optimizer='nadam')

model.save('keras_model.h5', include_optimizer=False)
```

but it fails as follows:
```
ValueError: Input 0 is incompatible with layer conv2d_1: expected ndim=4, found ndim=5
```

Could you adjust the minimal Python code aboce, in order to obtain a model that fails in frugally-deep, so I can reproduce the issue locally?

**Comment by katmatus** [comment]
Thank you for quick answer. Yes you are right, the problem is that your conv2d cannot process my input layer. I use functional model, this is part of it:
```
x_input = tf.keras.layers.Input((255,235, 1),name='x_input')
y_init =  tf.keras.layers.Input((52,2),name='y_init')

patches = tf.keras.layers.Lambda(extract_patches)([x_input,y_init])
conv2d1 = tf.keras.layers.Conv2D(32, (3,3), padding='same', activation='relu')(patches)


def extract_patches(inputs):
    list_patches = []
    for j in range(52):
        patch_one = tf.image.extract_glimpse(inputs[0], [26, 26], inputs[1][:, j, :], centered=False, 
        normalized=False, noise='zero')
        list_patches.append(patch_one)

    patches = tf.keras.backend.stack(list_patches,1)
    #reshape to (batch_size*52,26,26,1)
    return tf.keras.backend.reshape(patches,(-1,26,26,1))
```

**Comment by katmatus** [comment]
As input for conv2d during training I give shape (batch_size*52,26,26,1), where first dimension is batch_size  = batch_size*52. During prediction I have batch_size = 1, so shape is (1*52,26,26,1)

**Comment by katmatus** [comment]
I think the difference is that your conv2d can process only `batch_size = 1`

**Comment by Dobiasd** [comment]
Yes, you're right. A batch size other than 1 was never intended to be supported because in most usual model architectures it's just a performance thing, and making one single prediction with a model means everything just hast batch size 1.

But in your model is different because the custom layer converts a single prediction into some kind of batch in the middle of the model. I'll try to emulate what keras does in that case and get back to you with the results.

**Comment by katmatus** [comment]
Thank you very much, Tobias!

**Comment by Dobiasd** [comment]
Current I'm trying to better understand the implications of a model architecture like the one we are dealing with here.

So here is my attempt to re-create a model like that:

```python
# create_model.py
import numpy as np
import tensorflow as tf
from keras.layers import Input, Conv2D, Lambda
from keras.models import Model


def extract_patches(inputs):
    list_patches = []
    for j in range(3):
        list_patches.append(tf.image.extract_glimpse(inputs[0], [8, 8], inputs[1][:, j, :]))
    return tf.keras.backend.reshape(tf.keras.backend.stack(list_patches, 1), (-1, 8, 8, 1))


x_input = Input((16, 16, 1), name='x_input')
y_init = Input((3, 2), name='y_init')
patches = Lambda(extract_patches)([x_input, y_init])
conv2d1 = Conv2D(4, (3, 3), padding='same', activation='relu')(patches)

model = Model(inputs=[x_input, y_init], outputs=[conv2d1])
model.compile(loss='mean_squared_error', optimizer='nadam')
print(model.summary())
result = model.predict([np.zeros((1, 16, 16, 1)), np.zeros((1, 3, 2))])
print(result.shape)
model.save('keras_model.h5', include_optimizer=False)
```

When ran, it outputs the following summary:
```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
x_input (InputLayer)            (None, 16, 16, 1)    0                                            
__________________________________________________________________________________________________
y_init (InputLayer)             (None, 3, 2)         0                                            
__________________________________________________________________________________________________
lambda_1 (Lambda)               (None, 8, 8, 1)      0           x_input[0][0]                    
                                                                 y_init[0][0]                     
__________________________________________________________________________________________________
conv2d_1 (Conv2D)               (None, 8, 8, 4)      40          lambda_1[0][0]                   
==================================================================================================
Total params: 40
Trainable params: 40
Non-trainable params: 0
__________________________________________________________________________________________________
None
```

It's already interesting, the output shape is `(None, 8, 8, 4)`, but according to how I understand your report, it should be `(None, 3, 8, 8, 4)` in that case.

But at the end, the prediction step of this model fails:
```
  File "create_model.py", line 23, in <module>
    result = model.predict([np.zeros((1, 16, 16, 1)), np.zeros((1, 3, 2))])
  File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1169, in predict
    steps=steps)
  File "/usr/local/lib/python3.7/dist-packages/keras/engine/training_arrays.py", line 302, in predict_loop
    outs[i][batch_start:batch_end] = batch_out
ValueError: could not broadcast input array from shape (3,8,8,4) into shape (1,8,8,4
```

Could you help me out and modify the code above into something that works in Keras but then fails in frugally-deep?

**Comment by katmatus** [comment]
I changed `np.zeros` to `np.ones` in prediction code. Nothing is fails result.shape gives me `(3, 8, 8, 4)`. I use only tensorflow 2.0 and import keras like this `from tensorflow.python.keras.layers import Input, Conv2D, Lambda` . I have the same summary. Its ok like this ` None, 8, 8, 4`, cause this 3 goes in batch size, and when you use `batch_size = 1` it gives you shape `(1*3, 8, 8, 4)`

**Comment by katmatus** [comment]
You can download the model here [](https://drive.google.com/open?id=1-kq6qxvhNvmjbEJUWzKF61vZhkXDsm7Q)

**Comment by Dobiasd** [comment]
Thanks. I made the mentioned changes, and now I have:

```python3
# create_model.py
import numpy as np
import tensorflow as tf
from tensorflow.python.keras.layers import Input, Conv2D, Lambda
from tensorflow.python.keras.models import Model


def extract_patches(inputs):
    list_patches = []
    for j in range(3):
        list_patches.append(tf.image.extract_glimpse(inputs[0], [8, 8], inputs[1][:, j, :]))
    return tf.keras.backend.reshape(tf.keras.backend.stack(list_patches, 1), (-1, 8, 8, 1))


x_input = Input((16, 16, 1), name='x_input')
y_init = Input((3, 2), name='y_init')
patches = Lambda(extract_patches)([x_input, y_init])
conv2d1 = Conv2D(4, (3, 3), padding='same', activation='relu')(patches)

model = Model(inputs=[x_input, y_init], outputs=[conv2d1])
model.compile(loss='mean_squared_error', optimizer='nadam')
print(model.summary())
result = model.predict([np.ones((1, 16, 16, 1)), np.ones((1, 3, 2))])
print(result.shape)
model.save('keras_model.h5', include_optimizer=False)
```

Which also results in the following:
```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
x_input (InputLayer)            [(None, 16, 16, 1)]  0                                            
__________________________________________________________________________________________________
y_init (InputLayer)             [(None, 3, 2)]       0                                            
__________________________________________________________________________________________________
lambda (Lambda)                 (None, 8, 8, 1)      0           x_input[0][0]                    
                                                                 y_init[0][0]                     
__________________________________________________________________________________________________
conv2d (Conv2D)                 (None, 8, 8, 4)      40          lambda[0][0]                     
==================================================================================================
Total params: 40
Trainable params: 40
Non-trainable params: 0
__________________________________________________________________________________________________
  File "create_model.py", line 23, in <module>
    result = model.predict([np.ones((1, 16, 16, 1)), np.ones((1, 3, 2))])
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py", line 1078, in predict
    callbacks=callbacks)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training_arrays.py", line 370, in model_iteration
    aggregator.aggregate(batch_outs, batch_start, batch_end)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training_utils.py", line 169, in aggregate
    self.results[i][batch_start:batch_end] = batch_out
ValueError: could not broadcast input array from shape (3,8,8,4) into shape (1,8,8,4)
```

One (probably important) difference is the version I'm using:
```bash
$ python3 -c 'import keras; print(keras.__version__)'
Using TensorFlow backend.
2.2.4
```

```bash
$ python3 -c 'import tensorflow as tf; print(tf.__version__)'
1.14.0
```

Since TensorFlow 2 currently is still in the beta phase (`TensorFlow 2.0.0-beta1` according to [its releases page](https://github.com/tensorflow/tensorflow/releases)), frugally-deep does not yet officially support it.

Would it be feasible for you if we postpone this issue until TensorFlow has a stable release?

**Comment by katmatus** [comment]
Yes, I checked in Tensorflow 1.14 it has the same errors. The thing is I really want to predict my model in C++, probably you can give me advise how to make quick fix in your code to make it work?

**Comment by katmatus** [comment]
I made it work for you in tensorflow 1.14, it has to work in keras too
```
def extract_patches(inputs):
    list_patches = []
    for j in range(3):
        list_patches.append(tf.image.extract_glimpse(inputs[0], [8, 8], inputs[1][:, j, :]))
    return tf.keras.backend.reshape(tf.keras.backend.stack(list_patches, 1), (-1, 8, 8, 1))

def reshape(inputs):
    return  tf.keras.backend.reshape(inputs,(-1,3*8*8*4))


x_input = Input((16, 16, 1), name='x_input')
y_init = Input((3, 2), name='y_init')
patches = Lambda(extract_patches)([x_input, y_init])
conv2d1 = Conv2D(4, (3, 3), padding='same', activation='relu')(patches)
reshape = tf.keras.layers.Lambda(reshape)(conv2d1)
dense_1 = tf.keras.layers.Dense(10,activation='tanh')(reshape)
dense_2 = tf.keras.layers.Dense(6,activation=None)(dense_1)
prediction = tf.keras.layers.Reshape((3, 2))(dense_2)


model = Model(inputs=[x_input, y_init], outputs=[prediction])
model.compile(loss='mean_squared_error', optimizer='nadam')
print(model.summary())
result = model.predict([np.ones((1, 16, 16, 1)), np.ones((1, 3, 2))])
print(result.shape)
model.save('keras_model.h5', include_optimizer=False)
```

**Comment by Dobiasd** [comment]
Ah, that's helpful, thanks.

I can confirm that Keras with Tensorflow 1.14 actually accepts your minimal example, but I'm still not sure if what we are doing here is according to the Keras API specification, or if we just cleverly using a glitch in the implementation.

One clue is the fact, that changing:
```python3
model = Model(inputs=[x_input, y_init], outputs=[prediction])
```

to:
```python3
model = Model(inputs=[x_input, y_init], outputs=[prediction, conv2d1])
```

still results in:
```
ValueError: could not broadcast input array from shape (3,8,8,4) into shape (1,8,8,4)
```

Usually, it should always be possible to have one of the intermediate layer outputs also as an output of the whole model.

The [`Conv2D` documentation](https://keras.io/layers/convolutional/#conv2d) reads as follows:

> Input shape
>
> 4D tensor with shape: (batch, channels, rows, cols) if data_format is "channels_first" or 4D tensor with shape: (batch, rows, cols, channels) if data_format is "channels_last".
>
> Output shape
>
> 4D tensor with shape: (batch, filters, new_rows, new_cols) if data_format is "channels_first" or 4D tensor with shape: (batch, new_rows, new_cols, filters) if data_format is "channels_last". rows and cols values might have changed due to padding.

OK, but the model at hand manages to "trick" `Conv2D` into actually accepting `(3,8,8,4)` even though the actual batch size of the prediction call is 1 instead of 3.

Thus I assume the `Conv2D` layer just naively works as if it was called in a prediction step with batch size 3. At the same time I'm not sure what it will/should do in case the actual batch size is `!= 1`.

However, one way to hack emulation for this into frugally-deep might be a change in `fdeep::conv_2d_layer::apply_impl`, that does something like the following:
- Split the input tensor into multiple tensors according to `dim_4`.
- Run the actual call to `fdeep::internal::convolve` multiple times, once for each tensor. Re-concatenate the resulting tensors into one "batch" tensor, i.e., 3 `(8,8,4)` tensors back into one `(3,8,8,4)` tensor and return it.

Of course if you hack this into `fdeep/layers/conv_2d_layer.hpp`, you incur all the usual problems of privately maintaining such a fork of a library.

The reason why I'm hesitating to add this officially to frugally-deep is that I'm not sure if this really is intended Keras behavior. Thus I'm a bit scared about the can of worms we might open with adding such a thing. The only sane way might be to add proper batch support to frugally-deep as a whole.

Maybe there is another way to implement your use-case, one that works out of the box? :grimacing:

**Comment by katmatus** [comment]
Thank you very much for your advice. I will try to emulate batch in conv_2d_layer :) Probably it will not be so scary as you think, I will come back with results. I think that your light library could be very good analogue of a huge tensorflow, which is complicated to compile for C++, so you have to continue developing it to cover more cases. :)

**Comment by Dobiasd** [comment]
Looking forward to the results. :)

> Probably it will not be so scary as you think

Yeah, the quick implementation is probably quite simple and not very scary. And I'm totally fine if you do this in a fork that I don't have to maintain. :grimacing:

What I was thinking about is the long-term maintainability of the library in case this would be added officially. To me, the inconsistency introduced by such a hack just smells like a contact mine that could hamper development in the future, especially if we not yet even know if this a feature or just a glitch in Keras.

**Comment by katmatus** [comment]
There are a lot of opportunities now in Tensorflow 2.0. They removed huge part of Tensorflow and left only Keras support. So I think this behaviour will be normal for new version) I understand your fears, that adding batch size for Keras will lead to a big part of changes and  a lot of tests, but I think its possible :)
If you make a decision in future to extend it, probably I could help you:)

**Comment by Dobiasd** [comment]
Thanks for the help offered.

In general, I have nothing against adding batch support to frugally-deep, if it would bring a significant performance advantage. It's just that I'm not sure if having the batch size as just another tensor dimension is a good way to do it. Semantically I would prefer to separate it into a different data structure, but on the other hand, of course, we are not freely designing an independent framework here, but emulating Keras.

> So I think this behaviour will be normal for new version

That's the part I'm not sure enough about. The behavior of Keras the model at hand relies on still seems more like a glitch in Keras instead of a feature of it to me, because it seems to break my understanding of batching in a fundamental way. Batching should only be a performance thing to me, but not a change in the actual semantics of the computational graph.

Let's take the following example (using `tensorflow==2.0.0-beta1`):
```python3
import numpy as np
import tensorflow.keras.backend as k
from tensorflow.keras.layers import Input, Conv2D
from tensorflow.keras.models import Model

inputs = Input(shape=(8, 8, 6))
x = Conv2D(32, (3, 3))(inputs)
model = Model(inputs=inputs, outputs=x)
model.compile(loss='mean_squared_error', optimizer='nadam')
print(model.summary())
batch_size = 1
result = model.predict(np.ones((batch_size, 8, 8, 6)), batch_size=batch_size)
print(result.shape)
```
When ran with `batch_size = 1` we get an output tensor with shape `(1, 6, 6, 32)`. With `batch_size = 10` it becomes `(10, 6, 6, 32)`.

But a model like this (distilled version from the model we discussed above)
```python3
import numpy as np
import tensorflow.keras.backend as k
from tensorflow.keras.layers import Input, Conv2D, Lambda
from tensorflow.keras.models import Model


def custom_reshape(inputs):
    return k.reshape(inputs, (-1, 8, 8, 3))


inputs = Input(shape=(8, 8, 6))
x = Lambda(custom_reshape)(inputs)
x = Conv2D(32, (3, 3))(x)
model = Model(inputs=inputs, outputs=x)
model.compile(loss='mean_squared_error', optimizer='nadam')
print(model.summary())
batch_size = 1
result = model.predict(np.ones((batch_size, 8, 8, 6)), batch_size=batch_size)
print(result.shape)
```
outputs a `(2, 6, 6, 32)` Tensor. Changing to `batch_size = 10` should (intuitively) result in something with shape `(10, 2, 6, 6, 32)`, but of course it does not. Instead we get `(20, 6, 6, 32)`. I'm not sure how to even interpret such an output in a sane way. Manually splitting it up again seems somewhat weird.

Is the way you use `extract_glimpse` -> `stack` -> `reshape` -> `Conv2D` conventional? I mean, can you link to any references of others doing similar "crazy" stuff with the batch dimension?

**Comment by katmatus** [comment]
Yes, but this is not the conversation for

>  lambda layer tf.image

 issue :)
So to summarise, I successfully can run prediction of my model with small modifications. I added function `tensor5_from_dim4_slices` in `tensor5.hpp` to be able to concatenate tensors. I added loop in `pooling_2d_layer` and in `conv_2d_layer`. I could say that you have done a great job to create such a library!

**Comment by Dobiasd** [comment]
Thanks. :)

> I successfully can run prediction of my model with small modifications.

I'm glad you found a solution that works for you, even though it means you have to have your own fork of the library.

> Yes, but this is not the conversation for "lambda layer tf.image" issue :)

Probably, yes. In case you have some resources regarding this usage of the batch dimension, please let me know nonetheless. I'm interested in what kind of crazy stuff frugally-deep might need to support in the future. ^^
And if you like to continue to discuss in public, we can create a separate issue for that.

**Comment by Dobiasd** [comment]
To resolve the discussion about the batch dimension here: The behavior of TF 2.0 regarding this seems to [be a bug](https://groups.google.com/a/tensorflow.org/forum/#!topic/testing/txsgcR3cubQ). Thus I filed an [issue](https://github.com/tensorflow/tensorflow/issues/30474).

**Comment by Dobiasd** [comment]
Yay, the behavior is [fixed now in TF 2 and even has test coverage](https://github.com/tensorflow/tensorflow/commit/37fcf0a0e04b2014864936397c25e6c398135772). :)

**Comment by katmatus** [comment]
Hi so its bug. It doesn't influence my training, cause  it's working with
dense layers after even in tf 14. What could I say...library is library
with it's own logic, science is science...sometimes it goes out of the
box.:)
